### PR TITLE
Algorithm

### PR DESCRIPTION
**Pull request by CryptoSFor**

### CODE REVIEW

**Review by xartishuk — CHANGES_REQUESTED** [review]
this PR does 2 things: 
1. adds some general repository setup stuff 
2. adds business logic

split this into different PRs

**Comment by xartishuk** [comment]
this whole commit name would have probably fit into one string:
image
and even if it wouldn't fit, don't split it yourself

File: parentheses/algorithm.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// Package implements function BalancedString, that verifies if the given string is a balanceds sequence of brackets
+package parentheses
+
+// BalancedString verifies if the given string is a balanceds sequence of brackets
+func BalancedString(s string) bool {
+	list := []string{}
```
Comment (xartishuk) [inline parentheses/algorithm.go]: 1. rename `list` to `stack`
2. init your stack with capacity

File: parentheses/algorithm.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// Package implements function BalancedString, that verifies if the given string is a balanceds sequence of brackets
+package parentheses
+
+// BalancedString verifies if the given string is a balanceds sequence of brackets
```
Comment (xartishuk) [inline parentheses/algorithm.go]: typo

File: parentheses/algorithm.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// Package implements function BalancedString, that verifies if the given string is a balanceds sequence of brackets
```
Comment (xartishuk) [inline parentheses/algorithm.go]: typo

File: parentheses/algorithm.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		return true
+	}
+
+	if x1 == "[" && x2 == "]" {
+		return true
+	}
+
+	if x1 == "(" && x2 == ")" {
+		return true
+	}
```
Comment (xartishuk) [inline parentheses/algorithm.go]: this can be done with a single check using a map

File: parentheses/algorithm_test.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		{
+			name: "comparison of { and }",
+			args: args{x1: "}", x2: "{"},
+			want: false,
+		},
+		{
+			name: "comparison of empty string and [",
+			args: args{x1: "", x2: "]"},
+			want: false,
+		},
```
Comment (xartishuk) [inline parentheses/algorithm_test.go]: the names are wrong

File: parentheses/algorithm_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		{
+			name: "balanced1",
+			s:    "{([({}{}[][])])}",
+			want: true,
+		},
+		{
+			name: "balanced2",
+			s:    "效{[复]制___()+{仿效_}}制",
+			want: true,
+		},
```
Comment (xartishuk) [inline parentheses/algorithm_test.go]: if you want to test unicode, try `s: "(те[с]т){}",`
it should fail

**Review by xartishuk — CHANGES_REQUESTED** [review]

File: parentheses/algorithm.go (hunk 4 above)
Comment (CryptoSFor) [reply parentheses/algorithm.go]: Done

File: parentheses/algorithm.go (hunk 1 above)
Comment (CryptoSFor) [reply parentheses/algorithm.go]: Done

File: parentheses/algorithm.go (hunk 2 above)
Comment (CryptoSFor) [reply parentheses/algorithm.go]: Done

File: parentheses/algorithm.go (hunk 3 above)
Comment (CryptoSFor) [reply parentheses/algorithm.go]: Done

File: parentheses/algorithm_test.go (hunk 6 above)
Comment (CryptoSFor) [reply parentheses/algorithm_test.go]: actually, it works

File: parentheses/algorithm.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
+// Package implements function BalancedString, that verifies if the given string is a balanced sequence of brackets
+package parentheses
+
+// BalancedString verifies if the given string is a balanced sequence of brackets
+var brackets = map[string]string{"{": "}", "(": ")", "[": "]"}
+
+func BalancedString(s string) bool {
+	stack := make([]string, 0, 16)
```
Comment (xartishuk) [inline parentheses/algorithm.go]: why is cap =  16?

**Review by xartishuk — CHANGES_REQUESTED** [review]

File: parentheses/algorithm.go (hunk 7 above)
Comment (CryptoSFor) [reply parentheses/algorithm.go]: > 
What the capacity should be?

File: parentheses/algorithm_test.go (hunk 6 above)
Comment (xartishuk) [reply parentheses/algorithm_test.go]: ok than, i did some testing myself, you're right

File: parentheses/algorithm.go (hunk 7 above)
Comment (xartishuk) [reply parentheses/algorithm.go]: don't use magic number that don't have a reason to be there. If you don't have an explanation for 16, stick with the default.
Here you should base your `cap` on the size of `s`

Comment (CryptoSFor) [reply parentheses/algorithm.go]: Done

File: parentheses/algorithm.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+package parentheses
+
+// BalancedString verifies if the given string is a balanced sequence of brackets
+var brackets = map[string]string{"{": "}", "(": ")", "[": "]"}
+
+func BalancedString(s string) bool {
+	stack := make([]string, 0, len([]rune(s)))
+
+	for i := 0; i < len(s); i++ {
+		v := string(s[i])
```
Comment (xartishuk) [inline parentheses/algorithm.go]: you're counting the number of runes, but iterating over bytes

**Review by xartishuk — CHANGES_REQUESTED** [review]

File: parentheses/algorithm.go (hunk 8 above)
Comment (xartishuk) [reply parentheses/algorithm.go]: you should iterate over runes as well

Comment (xartishuk) [reply parentheses/algorithm.go]:
```go
for i := 0; i < len(s); i++ {
    v := string(s[i])
```
\+ it's much easier to just `range` over `string`

Comment (CryptoSFor) [reply parentheses/algorithm.go]: Done!

**Review by xartishuk — APPROVED** [review]

**Pull request merged**